feat: customizeable recording file name

JonasPammer opened the pull request:
Resolves apache/grails-core#14329

> Commit Message out-take:

>README.md text was copied with template from https://github.com/spockframework/spock/blob/master/docs/extensions.adoc#global-extensions
>
> I tried to get knowledge about this ServiceLoader paradigm from [grails-core code search](https://github.com/search?q=org%3Agrails+repo%3Agrails%2Fgrails-core+ServiceLoader.load&type=code) and [oracle doc](https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html)
>
> Although this whole META-INF is weird to me,
> I did not find it necessary to implement something complex like [spock's ConfigurationScriptLoader, which uses](https://github.com/spockframework/spock/blob/620f8734fb51c26cc9c8b8aade8d4dedc1532b94/spock-core/src/main/java/org/spockframework/runtime/ConfigurationScriptLoader.java#L34) (try to load env var path, file system, ..) into this just for this.

PR Appendix to last paragraph: Maybe the first one (env var) seems like a viable alternative/addition to this pattern, given that all other settings can be managed through `systemProperty 'grails.geb.recording.*` '*'``, i did not roll with using it instead as to first try to do it as per agreed upon https://github.com/grails/geb/issues/133#issuecomment-2654122264, and to possibly get validation if i did the ServiceLoader Pattern right ❤️ 💭

JonasPammer committed:
failing test

JonasPammer committed:
first try at ServiceLoader impl.

README.md text was copied with template from https://github.com/spockframework/spock/blob/master/docs/extensions.adoc#global-extensions

I tried to get knowledge about this ServiceLoader paradigm from https://github.com/search?q=org%3Agrails+repo%3Agrails%2Fgrails-core+ServiceLoader.load&type=code and https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html class javadoc

Although this whole META-INF is weird to me,
I did not find it necessary to implement something complex like https://github.com/spockframework/spock/blob/620f8734fb51c26cc9c8b8aade8d4dedc1532b94/spock-core/src/main/java/org/spockframework/runtime/ConfigurationScriptLoader.java#L34 (try to load env var path, file system, ..) into this just for this

# Conflicts:
#	README.md

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grails/geb?pullRequest=143) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grails/geb?pullRequest=143) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grails/geb?pullRequest=143) it.</sub>

JonasPammer committed:
chore: intellij

"left-shift cannot be applied to ..."

matrei commented on README.md:
```diff
@@ -94,6 +94,13 @@ By default, no test recording will be performed.  Various system properties exis
+in the class path (for example, `src/integration-test/resources`).
+As soon as these two conditions are satisfied, the class is automatically loaded and used instead of the classes default behaviour.
+You may use said [DefaultContainerGebTestDescription](./src/testFixtures/groovy/grails/plugin/geb/DefaultContainerGebTestDescription.groovy) 
+as a base class for your own implementation, and overwrite at minimum the `getTestId()` method to see a difference.
```
Suggestion:

To customize the naming convention for recording files:
1. Create a class that implements [ContainerGebTestDescription](./src/testFixtures/groovy/grails/plugin/geb/ContainerGebTestDescription.groovy).
2.  Specify its fully qualified class name in a `META-INF/services/grails.plugin.geb.ContainerGebTestDescription` file on the classpath (e.g., `src/integration-test/resources`).

Once both conditions are met, the class is automatically loaded and replaces the default behavior.  

You may extend the default implementation, [DefaultContainerGebTestDescription](./src/testFixtures/groovy/grails/plugin/geb/DefaultContainerGebTestDescription.groovy), as a base for your custom implementation. To modify the naming convention, override at least the `getTestId()` method.

JonasPammer replied:
Thanks, this is shorter and nicer, replaced

matrei reviewed (changes requested):
Awesome @JonasPammer, thank you for this PR!

I do have some concern for the tests, in that now all tests will use the custom implementation, and no tests will use the default implementation. If we create and use a `TestDescriptionServiceFactory` and call that instead of the `ServiceLoader` directly, we can modify the implementation programatically in `setupSpec()` and `cleanupSpec()`.

Also, I think we should target the `4.2.x` branch with this PR as we have not yet released a `4.2.0` version.

What do you think?

JonasPammer commented:
Totally! I saw other 2 projects in grails/spock also use a seperate class (i think there called loader), will implement thanks to your helpful guiding comments.

JonasPammer committed:
docs: much friendlier and easier english

JonasPammer committed:
refactor: factory pattern

JonasPammer committed:
chore: keep the META-INF for showcase purpose?

commit/file/folder can be removed, as documentation is clear

JonasPammer committed:
chore: make custom implementation clear

JonasPammer committed:
docs: document factory

JonasPammer committed:
docs: missing `

JonasPammer commented:
<details><summary>Force Push Explanation</summary>
<p>
[changed pr base branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request)

```
~\IdeaProjects\geb git:[geb-133-serviceloader]
git fetch upstream 4.2.x
From https://github.com/grails/geb
 * branch            4.2.x      -> FETCH_HEAD
~\IdeaProjects\geb git:[geb-133-serviceloader]
git fetch upstream 5.0.x
From https://github.com/grails/geb
 * branch            5.0.x      -> FETCH_HEAD
~\IdeaProjects\geb git:[geb-133-serviceloader]
git rebase --onto upstream/4.2.x upstream/5.0.x
Successfully rebased and updated refs/heads/geb-133-serviceloader.
~\IdeaProjects\geb git:[geb-133-serviceloader]
git push -f origin geb-133-serviceloader
Enumerating objects: 65, done.
Counting objects: 100% (65/65), done.
Delta compression using up to 12 threads
Compressing objects: 100% (30/30), done.
Writing objects: 100% (47/47), 5.58 KiB | 815.00 KiB/s, done.
Total 47 (delta 20), reused 0 (delta 0), pack-reused 0 (from 0)
remote: Resolving deltas: 100% (20/20), completed with 7 local objects.
To https://github.com/JonasPammer/geb.git
 + b2076f6...8186a9b geb-133-serviceloader -> geb-133-serviceloader (forced update)
```
</p>
</details>

JonasPammer committed:
chore: javadoc

JonasPammer committed:
suggestion: also add ContainerGebConfiguration ability

JonasPammer commented:
This PR (and, also seeing you there ;) gets me to also maybe open a PR in grails-docs for minor fixes found along the way, its fun

i am happy to drop some commits if desired, e.g. the last one - [[add ContainerGebConfiguration ability](https://github.com/grails/geb/pull/143/commits/a9045bf75d2b588d055716c53ec3cb008e3734d6)](https://github.com/grails/geb/pull/143/commits/a9045bf75d2b588d055716c53ec3cb008e3734d6)

matrei commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebTestDescriptionServiceFactory.groovy:
```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright 2024 original author or authors
```
2025

matrei commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebTestDescriptionServiceFactory.groovy:
```diff
@@ -0,0 +1,41 @@
+
+/**
+ * ServiceLoader (META-INF) Loader that allows programmatic overwriting.
+ */
```
`@Since 4.2`

matrei commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebTestDescriptionServiceFactory.groovy:
```diff
@@ -0,0 +1,41 @@
+ * ServiceLoader (META-INF) Loader that allows programmatic overwriting.
+ */
+@CompileStatic
+class ContainerGebTestDescriptionServiceFactory {
```
Is `ContainerGebTestDescriptionServiceLoader` or `ContainerGebTestDescriptionServiceResolver` a better name?

JonasPammer replied:
after looking at https://github.com/search?q=org%3Agrails+ServiceLoader.load&type=code&rgh-link-date=2025-02-21T21%3A21%3A37Z i went with Loader because of grails-datastore-core/src/main/groovy/org/grails/datastore/mapping/services/SoftServiceLoader.java, and because its not a real world factory (its more like a fancy singleton class) ;)

matrei commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebTestDescriptionServiceFactory.groovy:
```diff
@@ -0,0 +1,41 @@
+ */
+@CompileStatic
+class ContainerGebTestDescriptionServiceFactory {
+    private static ContainerGebTestDescription implementation = null
```
Blank line after class definition for better readability?
Does the name `instance` better reflect what this field is?

jdaugherty replied:
the assignment to null is unnecessary

matrei commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebTestDescriptionServiceFactory.groovy:
```diff
@@ -0,0 +1,41 @@
+class ContainerGebTestDescriptionServiceFactory {
+    private static ContainerGebTestDescription implementation = null
+
+    static ContainerGebTestDescription getImplementation() {
```
`getInstance()`?

matrei commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebTestDescriptionServiceFactory.groovy:
```diff
@@ -0,0 +1,41 @@
+    private static ContainerGebTestDescription implementation = null
+
+    static ContainerGebTestDescription getImplementation() {
+        return implementation ?: ServiceLoader.load(ContainerGebTestDescription).findFirst().orElse(new DefaultContainerGebTestDescription())
```
Break long line for readability?

matrei commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebTestDescriptionServiceFactory.groovy:
```diff
@@ -0,0 +1,41 @@
+        return implementation ?: ServiceLoader.load(ContainerGebTestDescription).findFirst().orElse(new DefaultContainerGebTestDescription())
+    }
+
+    static void setImplementation(ContainerGebTestDescription implementation) {
```
`setInstance(...)`?

matrei commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebTestDescriptionServiceFactory.groovy:
```diff
@@ -0,0 +1,41 @@
+    /**
+     * Class must have a zero-argument constructor (ServiceLoader Requirement).
+     */
+    static void setImplementation(Class<? extends ContainerGebTestDescription> implementation) {
```
`setImplementationClass(Class<? extends ContainerGebTestDescription> clazz)`?

JonasPammer replied:
I like Java Overloads, but could do

matrei commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebConfiguration.groovy:
```diff
@@ -52,4 +53,11 @@ import java.lang.annotation.Target
+    /**
+     * The {@link org.testcontainers.lifecycle.TestDescription} implementation to use for this class.
+     * <p> {@link NullContainerGebTestDescription} results in the
+     *     {@link ContainerGebTestDescriptionServiceFactory last set} implementation being used.
```
Should ` last set` be outside the `@link`?

JonasPammer replied:
for once, this was on purpose :)

JonasPammer replied:
![image](https://github.com/user-attachments/assets/257b101a-6057-4b0e-8fed-49e97ffbc5f1)
wanted to hide it, as reading this long of a java pattern class name didnt feel nice

jdaugherty commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebConfiguration.groovy:
```diff
@@ -34,6 +34,7 @@ import java.lang.annotation.Target
 
     static final String DEFAULT_HOSTNAME_FROM_CONTAINER = GenericContainer.INTERNAL_HOST_HOSTNAME
     static final String DEFAULT_PROTOCOL = 'http'
+    static final Class<? extends ContainerGebTestDescription> DEFAULT_TEST_DESCRIPTION = NullContainerGebTestDescription.class
```
You can drop the `class` since we're in groovy.

jdaugherty commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebTestDescription.groovy:
```diff
@@ -20,26 +20,10 @@ import org.spockframework.runtime.model.IterationInfo
-    }
+interface ContainerGebTestDescription extends TestDescription {
+    IterationInfo getTestInfo()
+    void setTestInfo(IterationInfo testInfo)
```
The junit 5 api already has a class called TestInfo, which could cause some confusion here.  Can we please rename this to match the spock naming so people know where to look in the spock documentation?

jdaugherty commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebConfiguration.groovy:
```diff
@@ -52,4 +53,11 @@ import java.lang.annotation.Target
+     * <p> {@link NullContainerGebTestDescription} results in the
+     *     {@link ContainerGebTestDescriptionServiceFactory last set} implementation being used.
+     */
+    Class<? extends ContainerGebTestDescription> testDescription() default NullContainerGebTestDescription.class
```
Since this groovy, the `.class` is unnecessary

jdaugherty commented on spock-container-test-app/src/integration-test/groovy/org/demo/spock/ContainerGebTestDescriptionImpl.groovy:
```diff
@@ -0,0 +1,10 @@
+
+import grails.plugin.geb.DefaultContainerGebTestDescription
+
+class ContainerGebTestDescriptionImpl extends DefaultContainerGebTestDescription {
```
`@CompileStatic`

jdaugherty commented on spock-container-test-app/src/integration-test/groovy/org/demo/spock/ContainerGebTestDescriptionImpl.groovy:
```diff
@@ -0,0 +1,10 @@
+class ContainerGebTestDescriptionImpl extends DefaultContainerGebTestDescription {
+    @Override
+    String getTestId() {
+        return "Custom_" + testInfo.feature.spec.displayName
```
Instead of using string concatenation, can we use a GString:

       return "Custom_${testInfo.feature.spec.displayName}" as String

jdaugherty commented:
@JonasPammer Thank you for taking the time on this.  I had only minor comments / cleanup.

JonasPammer committed:
code review comments

JonasPammer committed:
chore: update year in license header

as done by https://github.com/grails/geb/pull/148/commits/947eb0ae8ef59c0e1a47ee01173664ad8248843d

JonasPammer committed:
feat: allow implementation to access errorInfo

although default doesnt use it, as to not change default behaviour with this PR, one may want to?

JonasPammer committed:
refactor: ensure getter of implementation is used, to open up mocking/faking possibility?

just an idea

matrei commented on spock-container-test-app/src/integration-test/groovy/org/demo/spock/ContainerGebTestDescriptionSpec.groovy:
```diff
@@ -0,0 +1,33 @@
+        param_3 << Integer.MAX_VALUE
+    }
+
+    def cleanupSpec(){
+        ContainerGebTestDescriptionServiceFactory.setImplementation(null)
+    }
```
Move method to be below `setupSpec()`?

JonasPammer replied:
OK

matrei commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebTestDescription.groovy:
```diff
@@ -16,30 +16,38 @@
-                testInfo.displayName != testInfo.feature.displayName ? testInfo.iterationIndex : null
-        ].findAll(/* Remove nulls */).join(' ')
+interface ContainerGebTestDescription extends TestDescription {
+    IterationInfo getIterationInfo()
```
Blank line after class definition and between methods for readability?

matrei commented on src/testFixtures/groovy/grails/plugin/geb/WebDriverContainerHolder.groovy:
```diff
@@ -117,6 +117,9 @@ class WebDriverContainerHolder {
             configObject.reportsDir = grailsGebSettings.reportingDirectory
             configObject.reporter = (invocation.sharedInstance as ContainerGebSpec).createReporter()
         }
+        if (currentConfiguration.testDescription != NullContainerGebTestDescription.class) {
```
No need for `.class` in Groovy.

matrei commented on spock-container-test-app/src/integration-test/groovy/org/demo/spock/ContainerGebTestDescriptionAnnotationSpec.groovy:
```diff
@@ -0,0 +1,21 @@
+import grails.testing.mixin.integration.Integration
+
+@Integration
+@ContainerGebConfiguration(testDescription = ContainerGebTestDescriptionImpl.class)
```
No need for `.class` in Groovy.

matrei commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebTestDescription.groovy:
```diff
@@ -16,30 +16,38 @@
+     * You could overwrite {@link ErrorInfo#getException()} to mimic that an Exception occurred, resulting in a recording file
+     * if {@link org.testcontainers.containers.BrowserWebDriverContainer#recordingMode} is RECORD_FAILING.
+     */
+    @Nullable
```
I don't think we should use `javax.annotation.Nullable`. It is a legacy annotation.

JonasPammer replied:
Yeah, i never know which Nullable to take, i just took the one with java in front. i think there was also a spock and spring one, but maybe ill make it just a comment

JonasPammer replied:
Went with the Spring one, as has lots of counterparts an big javadoc comment, but first thought of making it a comment in own line / paragraph

matrei replied:
This is fine, but we should add:
```groovy
testFixturesImplementation 'org.springframework:spring-core', {
    // impl: @Nullable(runtime)
}
```
below `selenium-support` in `build.gradle` to show that this is a direct dependency.

matrei commented on spock-container-test-app/src/integration-test/groovy/org/demo/spock/ContainerGebTestDescriptionSpec.groovy:
```diff
@@ -0,0 +1,33 @@
+        true
+
+        where:
+        data_driven_testing_param_1 << ["very long text lorem ipsum dolor sit amet lorem ipsum dolor sit amet lorem ipsum dolor sit amet Liskov Substitution"]
+        param_2 << ["If an implementation is hard to explain, it is an established design pattern."]
```
Use single quotes for Strings?

matrei reviewed (commented):
Excellent! I have some minor suggestions/questions.

JonasPammer commented:
thanks for helping me improve, will implement changes sometime tomorrow

JonasPammer committed:
chore: code review <3

JonasPammer committed:
chore: code review

JonasPammer commented:
left 2 comments open ended to either be resolved or answered with that we should do it the way written

matrei commented on README.md:
```diff
@@ -76,6 +76,20 @@ By default, no test recording will be performed.  Here are the system properties
+2. Specify its fully qualified class name in a `META-INF/services/grails.plugin.geb.ContainerGebTestDescription` file
+   on the classpath (e.g., `src/integration-test/resources`).
+
+Once both conditions are met, the class is automatically loaded and replaces the default behavior. +
```
Remove `+` at the end of the line?

matrei commented on spock-container-test-app/src/integration-test/groovy/org/demo/spock/ContainerGebTestDescriptionAnnotationSpec.groovy:
```diff
@@ -0,0 +1,21 @@
+        true
+
+        where:
+        data_driven_testing_param_1 << ["very long text lorem ipsum dolor sit amet lorem ipsum dolor sit amet lorem ipsum dolor sit amet Liskov Substitution"]
```
Single quotes for Strings?

matrei commented on spock-container-test-app/src/integration-test/groovy/org/demo/spock/ContainerGebTestDescriptionAnnotationSpec.groovy:
```diff
@@ -0,0 +1,21 @@
+@ContainerGebConfiguration(testDescription = ContainerGebTestDescriptionImpl)
+class ContainerGebTestDescriptionAnnotationSpec extends ContainerGebSpec {
+
+    void 'should not raise a FileSystemException on recording save by loading custom implementation of TODO'(String data_driven_testing_param_1) {
```
Is `TODO` supposed to be in the title?

JonasPammer replied:
ah nah, i hadnt thought of a name there yet, thanks

matrei commented on spock-container-test-app/src/integration-test/groovy/org/demo/spock/ContainerGebTestDescriptionSpec.groovy:
```diff
@@ -0,0 +1,33 @@
+    /**
+     * "The filename, directory name, or volume label syntax is incorrect" (too long)
+     */
+    void 'should not raise a FileSystemException on recording save by loading custom implementation of TODO'(String data_driven_testing_param_1, String param_2, int param_3) {
```
Is `TODO` supposed to be in the title?

matrei commented on spock-container-test-app/src/integration-test/resources/META-INF/services/grails.plugin.geb.ContainerGebTestDescription:
```diff
@@ -0,0 +1 @@
+grails.plugin.geb.DefaultContainerGebTestDescription
```
I'm not sure we should keep this file. The default is loaded anyway, so keeping it might actually be misleading that it is mandatory.

matrei commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebTestDescription.groovy:
```diff
@@ -16,30 +16,38 @@
 
 /**
- * Implements {@link org.testcontainers.lifecycle.TestDescription} to customize recording names.
+ * java.util.ServiceLoader Compatible Interface ("must have a zero-argument constructor")
```
Suggestion:
```
/**
 * An extension of {@link org.testcontainers.lifecycle.TestDescription} that allows customization of recording file names.
 * <p>
 * Implementations must provide a zero-argument constructor to ensure compatibility with {@link java.util.ServiceLoader}.
```

matrei commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebTestDescription.groovy:
```diff
@@ -16,30 +16,38 @@
+     */
+    @Nullable
+    ErrorInfo getErrorInfo()
+    /**
```
Blank line above new method?

matrei commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebTestDescription.groovy:
```diff
@@ -16,30 +16,38 @@
-    String testId
-    String filesystemFriendlyName
+    IterationInfo getIterationInfo()
+    /**
```
Blank line above new method?

matrei commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebTestDescriptionServiceLoader.groovy:
```diff
@@ -0,0 +1,46 @@
+import groovy.transform.CompileStatic
+
+/**
+ * ServiceLoader (META-INF) Loader that allows programmatic overwriting.
```
Suggestion:
```
/**
 * ServiceLoader (META-INF) for {@link ContainerGebTestDescription} that allows programmatic overwriting.
```

matrei commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebTestDescriptionServiceLoader.groovy:
```diff
@@ -0,0 +1,46 @@
+    static void setInstance(ContainerGebTestDescription instance) {
+        this.instance = instance
+    }
+    /**
```
Blank line above new method?

matrei commented on src/testFixtures/groovy/grails/plugin/geb/DefaultContainerGebTestDescription.groovy:
```diff
@@ -0,0 +1,56 @@
+ * Used by {@link ContainerGebConfiguration#testDescription()} interface to represent a null value.
+ */
+class NullContainerGebTestDescription extends DefaultContainerGebTestDescription {
+
```
Remove empty line?

matrei commented on src/testFixtures/groovy/grails/plugin/geb/GebRecordingTestListener.groovy:
```diff
@@ -41,9 +41,13 @@ class GebRecordingTestListener extends AbstractRunListener {
 
     @Override
     void afterIteration(IterationInfo iteration) {
+        final ContainerGebTestDescription testContainerDescription = ContainerGebTestDescriptionServiceLoader.getInstance()
```
Do we need `final` here?

matrei reviewed (commented):
Nice! Just some follow up comments/suggestions.

JonasPammer committed:
code review alterations

JonasPammer commented:
👍

matrei commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebTestDescriptionServiceLoader.groovy:
```diff
@@ -0,0 +1,47 @@
+
+    private static ContainerGebTestDescription instance
+
+    static ContainerGebTestDescription getInstance() {
+        if (instance != null) return instance
+        return ServiceLoader.load(ContainerGebTestDescription).findFirst()
+                .orElse(new DefaultContainerGebTestDescription())
+    }
```
(Sorry, this comment seem to have vanished from my previous review)
Can be simplified:
```groovy
static ContainerGebTestDescription getInstance() {
    instance ?: ServiceLoader.load(ContainerGebTestDescription)
            .findFirst()
            .orElse(new DefaultContainerGebTestDescription())
}
```

matrei commented on spock-container-test-app/src/integration-test/groovy/org/demo/spock/ContainerGebTestDescriptionImpl.groovy:
```diff
@@ -0,0 +1,12 @@
+import groovy.transform.CompileStatic
+
+@CompileStatic
+class ContainerGebTestDescriptionImpl extends DefaultContainerGebTestDescription {
```
Blank line after class declaration for readability?

JonasPammer committed:
code review alterations

JonasPammer committed:
move *Loader to grails.geb.serviceloader. as per https://github.com/grails/geb/pull/146#issuecomment-2686107413

intellij refactor + Spec needed manually

JonasPammer commented:
last commit can be dropped, i wanted to hold a grail to the actual code of your alls which this plugin is about to not be bloated by java design patterns necessities of 4 classes

matrei commented:
Thanks @JonasPammer, I think we can drop the last commit for now (it might actually be a breaking change in 4.x).
After that I'm happy with this PR and will gladly approve.

JonasPammer committed:
dropped https://github.com/grails/geb/pull/143/commits/058d3b22b4522f4d60e52789a5958189d9dbcfef, but kept README fix

matrei commented:
Talking of breaking changes. Changing `ContainerGebTestDescription` to an interface is actually a breaking change. @jdaugherty What do you think, should we target `5.0.x` with this PR instead?

JonasPammer commented:
if its just about the namespacing, i could make `ContainerGebTestDescription` be the current `Default....`, and transform the interface class name to `IContainer...`, if the `I` pattern is ok

jdaugherty commented:
> Talking of breaking changes. Changing `ContainerGebTestDescription` to an interface is actually a breaking change. @jdaugherty What do you think, should we target `5.0.x` with this PR instead?

It was internal only so I think it's ok in this case.

JonasPammer commented:
given [![image](https://github.com/user-attachments/assets/fefc7a94-0f0e-4cf2-ae1c-06c4d97dd0b7)](https://github.com/grails/geb/pull/143#issuecomment-2688315561) i will go ahead with the `Default..` Naming scheme in #146

matrei commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebTestDescription.groovy:
```diff
@@ -16,30 +16,43 @@
+ * Implementations must provide a zero-argument constructor to ensure compatibility with {@link java.util.ServiceLoader}.
  *
- * @author James Daugherty
- * @since 4.1
```
`@since 4.2`?

JonasPammer committed:
oki

matrei commented:
@JonasPammer Should we go in the direction of a more general service loader instead, because it looks like really usable extension point mechanism that we can useful for more cases: https://github.com/grails/geb/pull/146#issuecomment-2690523380

JonasPammer commented:
set do draft as per above comment linked PR design decision WIP

JonasPammer committed:
refactor to last PRs

JonasPammer committed:
rebase fixes

JonasPammer commented:
as per https://github.com/apache/grails-geb/issues/132#issuecomment-2816889221 i understand when this will be closed i will reopen a new version in core

jdaugherty commented:
@JonasPammer you should be safe to open this against grails core now.  

grails-geb is the source location in core
grails-test-examples/geb is the test app
